src/util/include/util/UnionOfRectangles.h

```diff
@@ -0,0 +1,52 @@
+/*
+ * Xournal++
+ *
+ * A rectangle data structure over a numeric data type.
+ *
+ * @author Xournal++ Team
+ * https://github.com/xournalpp/xournalpp
+ *
+ * @license GNU GPLv2 or later
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <vector>
+
+#include "util/Rectangle.h"
+
+
+template <class T>
+class UnionOfRectangles final {
+public:
+    UnionOfRectangles() = default;
+    using RectangleType = xoj::util::Rectangle<T>;
+
+    void addRectangle(const RectangleType& rect) {
+        const RectangleType* last_united = &rect;
+        for (auto& old_rect: data) {
+            if (old_rect.intersects(*last_united)) {
+                old_rect.unite(*last_united);
+                last_united = &old_rect;
+            }
+        }
+        if (last_united == &rect) {
+            data.emplace_back(rect);
+            return;
+        }
+        // Move the addition to the front we want to keep it
+        std::swap(data.front(), *const_cast<RectangleType*>(last_united));
+        data.erase(std::remove_if(std::next(begin(data)), end(data),
+                                  [=](auto const& other_rect) {
+                                      // intersecting rects are united in data.front();
+                                      //                                       other_rect.contains(rect);
+                                      return other_rect.x <= rect.x && other_rect.y <= rect.y &&
+                                             other_rect.x + other_rect.width >= rect.x + rect.width &&
+                                             other_rect.y + other_rect.height >= rect.y + rect.height;
+                                  }),
+                   end(data));
+    }
```

Comment:
This proposal does not pass `ctest`.
There needs to be several passes because as the `last_united` grows, it may start intersecting (without containing) another rectangle that we already tested. See the test case in test/unit_tests/util/RectangleTest.cpp.

Let me try to find a better version of that using more std::algorithm than my original code.

Reply:
I made a new proposal that passes the `ctest` and is more clear, I think.

Reply:
Good you made those tests. I am just not sure, if it would be faster to only allow one repaint rectangle. Currently, we will have at worse a runtime of P(S) (power set), which is extremely dangerous. Maybe a more heuristic but not so correct approach is better. 

(I would leave a comment that this can be a pitfall, we can fix that later if required.)

Reply:
Yeah, it's really not clear how much we gain from being greedy on the rerendered areas. Clearly, there's the MacOs hardware acceleration issue playing a part in those choices.

Maybe it'd be interesting to get how many rectangles are (usually) rerendered in by a render job (see `PageView` L700+ and `RenderJob`). If this is close to 1, then this class can be dropped altogether, and replaced with a single rectangle to which we add everything.

Reply:
I made some measurements. The average number of rectangles rendered by a single RenderJob run is around 1.01. (I once reached 1.02 :-). So yeah, we should just unite all rectangles.

Reply:
I removed UnionOfRectangles in the last commit.